sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1618,23 +1619,198 @@ case class ArrayBinarySearch(array: Expression, value: Expression)
+
+  private def unboxArgumentsAndInputTypes(): (Seq[Expression], Seq[AbstractDataType]) =
+    elementType match {
+      case ByteType =>
+        (Seq(Literal(foldableByteArrayData, foldableByteArrayDataObjectType), value),
+          Seq(foldableByteArrayDataObjectType, elementType))
+      case ShortType =>
+        (Seq(Literal(foldableShortArrayData, foldableShortArrayDataObjectType), value),
+          Seq(foldableShortArrayDataObjectType, elementType))
+      case IntegerType =>
+        (Seq(Literal(foldableIntArrayData, foldableIntArrayDataObjectType), value),
+          Seq(foldableIntArrayDataObjectType, elementType))
+      case LongType =>
+        (Seq(Literal(foldableLongArrayData, foldableLongArrayDataObjectType), value),
+          Seq(foldableLongArrayDataObjectType, elementType))
+      case FloatType =>
+        (Seq(Literal(foldableFloatArrayData, foldableFloatArrayDataObjectType), value),
+          Seq(foldableFloatArrayDataObjectType, elementType))
+      case DoubleType =>
+        (Seq(Literal(foldableDoubleArrayData, foldableDoubleArrayDataObjectType), value),
+          Seq(foldableDoubleArrayDataObjectType, elementType))
+  }
+
+  private def boxedArgumentsAndInputTypes(): (Seq[Expression], Seq[AbstractDataType]) =
+    elementType match {
+      case BooleanType =>
+        (Seq(Literal(foldableBoxedBooleanArrayData, foldableBoxedBooleanArrayDataObjectType),
+          value),
+          Seq(foldableBoxedBooleanArrayDataObjectType, elementType))
+      case ByteType =>
+        (Seq(Literal(foldableBoxedByteArrayData, foldableBoxedByteArrayDataObjectType), value),
+          Seq(foldableBoxedByteArrayDataObjectType, elementType))
+      case ShortType =>
+        (Seq(Literal(foldableBoxedShortArrayData, foldableBoxedShortArrayDataObjectType), value),
+          Seq(foldableBoxedShortArrayDataObjectType, elementType))
+      case IntegerType =>
+        (Seq(Literal(foldableBoxedIntArrayData, foldableBoxedIntArrayDataObjectType), value),
+          Seq(foldableBoxedIntArrayDataObjectType, elementType))
+      case LongType =>
+        (Seq(Literal(foldableBoxedLongArrayData, foldableBoxedLongArrayDataObjectType), value),
+          Seq(foldableBoxedLongArrayDataObjectType, elementType))
+      case FloatType =>
+        (Seq(Literal(foldableBoxedFloatArrayData, foldableBoxedFloatArrayDataObjectType), value),
+          Seq(foldableBoxedFloatArrayDataObjectType, elementType))
+      case DoubleType =>
+        (Seq(Literal(foldableBoxedDoubleArrayData, foldableBoxedDoubleArrayDataObjectType), value),
+          Seq(foldableBoxedDoubleArrayDataObjectType, elementType))
+    }
+
+  override def replacement: Expression = {
     if (canPerformFastBinarySearch) {
-      StaticInvoke(
-        classOf[ArrayExpressionUtils],
-        IntegerType,
-        "binarySearch",
-        Seq(array, value),
-        inputTypes)
+      if (arrayIsFoldable) {
```

Comment:
I'm a bit confused. What's the difference between foldable and non-foldable arrays regarding this optimization?

Reply:
In the case where the `array` is `foldable`, after optimization, the array `only` needs to be `toArray` `once` instead of toArray `every time`.